mlflow/evaluation.py

```diff
@@ -0,0 +1,424 @@
+            }
+            for artifact_name, artifact in self.artifacts.items()
+        }
+        with open(os.path.join(path, "artifacts_metadata.json"), "w") as fp:
+            json.dump(artifacts_metadata, fp)
+
+        for artifact_name, artifact in self.artifacts.items():
+            artifact.save_content_to_file(artifact.content, os.path.join(path, artifact_name))
+
+    @property
+    def metrics(self) -> EvaluationMetrics:
+        """
+        A dictionary mapping scalar metric names to scalar metric values
+        """
+        return self._metrics
+
+    @property
+    def artifacts(self) -> Dict[str, EvaluationArtifact]:
+        """
+        A dictionary mapping standardized artifact names (e.g. "roc_data") to
+        artifact content and location information
+        """
+        return self._artifacts
+
+
+class EvaluationDataset:
+    """
+    Represents an input dataset for model evaluation. This is intended for
+    use with the `mlflow.evaluate()`API.
+    """
+
+    NUM_SAMPLE_ROWS_FOR_HASH = 5
+
+    def __init__(self, data, labels, name=None, path=None):
+        """
+        :param data: One of the following:
+         - A numpy array or list of evaluation features, excluding labels.
+         - A Pandas DataFrame, or the path to a serialized DataFrame,
+           containing evaluation features and labels. All columns will be regarded as feature
+           columns except the "labels" column.
+
+        :param labels: One of the following:
+         - A numpy array or list of evaluation labels, if `data` is also a numpy array or list.
+         - The string name of a column from `data` that contains evaluation labels, if `data`
+           is a DataFrame.
+
+        :param name: (Optional) The name of the dataset (must not contain ").
+
+        :param path: (Optional) the path to a serialized DataFrame
+          (e.g. a delta table, parquet file)
+        """
+        self.user_specified_name = name
+        self.data = data
+        self.labels = labels
+        self.path = path
+        self._hash = None
+
+    def extract_features_and_labels(self):
+        if isinstance(self.data, np.ndarray):
+            return self.data, self.labels
+        elif isinstance(self.data, pd.DataFrame):
+            feature_cols = [x for x in self.data.columns if x != self.labels]
+            return self.data[feature_cols], self.data[self.labels]
+        else:
+            raise ValueError(f'Unsupported data type: {type(self.data)}')
+
+    @staticmethod
+    def _gen_md5_for_arraylike_obj(md5_gen, data):
+        md5_gen.update(pickle.dumps(len(data)))
+        if len(data) < EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH * 2:
+            md5_gen.update(pickle.dumps(data))
+        else:
+            md5_gen.update(pickle.dumps(data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]))
+            md5_gen.update(pickle.dumps(data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]))
```

Comment:
If we use pickle, this hash will change across Python versions, right? It could also change depending on other aspects of the current environment, e.g. the point in the callstack where `mlflow.evaluate()` is called. Could we instead use something like https://pandas.pydata.org/docs/reference/api/pandas.util.hash_pandas_object.html for pandas dataframes and https://stackoverflow.com/a/5386770/11952869 for numpy arrays? Feel free to do your own research here.

Reply:
The function https://pandas.pydata.org/docs/reference/api/pandas.util.hash_pandas_object.html will return a pandas Series object which contains a list of uint values, which is not convenient. I rethink on this, I suggest using this:
```
    def _array_like_obj_to_bytes(data):
        if isinstance(data, pd.DataFrame):
            return data.to_numpy().tobytes()
        elif isinstance(data, np.ndarray):
            return data.tobytes()
        elif isinstance(data, list):
            return np.array(data).tobytes()
```
This helper method convert pandas dataframe/numpy array/list into a deterministic bytes array, and it is independent with python version/platform.